sympy/functions/combinatorial/numbers.py

```diff
@@ -2175,6 +2175,149 @@ def eval(cls, n):
+#                                                                             #
+#                         Schoeder numbers                                    #
+#                                                                             #
+#-----------------------------------------------------------------------------#
+
+
+class schroeder(Function):
+    r"""
+    In mathematics, the Schroeder number,
+    also called a large Schroeder number or
+    big Schroeder number, describes the number of
+    lattice paths from the southwest corner $(0, 0)$
+    of an $n\times n$ grid to the northeast corner $(n, n)$ using only
+    single steps north, $(0, 1)$ northeast, $(1, 1)$ or east, $(1, 0)$
+    that do not rise above the SW-NE diagonal. The nth
+    Schroeder number is the number of Schroeder paths
+    of length n. The Schroeder numbers  were named
+    after the German mathematician Ernst Schroeder.
+
+    Schroeder numbers are the integer sequences defined by the
+    initial terms `S_0 = 1`, `S_1 = 2` and the two-term recurrence relation
+    `S_n = \frac{6*n - 3}{n + 1} * S_{n-1} - \frac{n - 2}{n + 1} * S_{n-2}`.
+
+
+    Examples
+    ========
+
+    >>> from sympy import schroeder
+
+    >>> schroeder(2)
+    2
+    >>> schroeder(5)
+    90
+    >>> schroeder.is_schroeder(2)
+    True
+    >>> schroeder.is_schroeder(7)
+    False
+    >>> schroeder.find_first_n_schroeder_numbers(1)
+    [1]
+    >>> schroeder.find_first_n_schroeder_numbers(10)
+    [1, 2, 6, 22, 90, 394, 1806, 8558, 41586, 206098]
+    >>> schroeder.find_schroeder_numbers_in_range(1, 10)
+    [1, 2, 6]
+    >>> schroeder.find_schroeder_numbers_in_range(2, 400)
+    [2, 6, 22, 90, 394]
+
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Schr%C3%B6der_number
+    .. [2] https://mathworld.wolfram.com/SchroederNumber.html
+
+    """
+
+    @staticmethod
+    def is_schroeder(n):
+        try:
+            n = as_int(n)
+        except ValueError:
+            False
+        if n > 0:
+            if n in (1, 2):
+                return True
+
+            sn1 = 1
+            sn = 2
+            i = 2
+            while sn < n:
+                b = ((6*i - 3)*sn - (i - 2)*sn1) // (i + 1)
+                sn1 = sn
+                sn = b
+                i = i+1
+
+            return sn == n
+
+        else:
+            return False
+
+    @staticmethod
+    def find_first_n_schroeder_numbers(n):
+        try:
+            n = as_int(n)
+        except ValueError:
+            raise ValueError('The provided number must be a positive integer')
+        if n < 0:
+            raise ValueError('The provided number must be a positive integer')
+        if n == 1:
+            return [1]
+        schroeder = [1, 2]
+        sn1 = 2
+        sn = 6
+        i = 3
+        while i <= n:
+            schroeder.append(sn)
+            b = ((6*i - 3)*sn - (i - 2)*sn1) // (i + 1)
```

Comment:
This code for iterating over the Schroeder numbers is repeated in two different methods and this formula is also repeated in the `_schroeder` function. It should be possible to factor out common code. An iterator that yields the schroeder numbers could be used to implement at least these two methods without duplicating the code.

Reply:
I have added a iterator function to iterate over schroeder numbers but I'm not sure whether it is the correct way.